### PR TITLE
cmd/snap-confine: handle death of helper process

### PR DESCRIPTION
**Pull request by zyga**
When SIGPIPE is delivered the snap-confine process it would involuntary
terminate. If snap-confine was a part of a startup chain of a systemd
service then systemd would consider the exit as _clean_, as documented
by systemd.service(5), even though the actual snap application was never
invoked. This patch corrects this in the following way.

When the snap-confine helper process terminates due to SIGARLM-based
timeout mechanism the parent would receive SIGPIPE on the attempt to
communicate with the helper.

The disposition of SIGPIPE is set to SIG_IGN, and only in the scope where we
are interacting with the helper process and are possibly using pipes for 
communication. This ensures that the helper dies for whatever reason the
parent can just deal with EPIPE (and exiting with a proper error code) instead
of dying because of SIGPIPE.

The patch also contains a regression test, for easier review:

Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

### CODE REVIEW

**Comment by zyga** [comment]
The test is flawed because fexecve doesn't support interpreted scripts the way we use them.

**Comment by zyga** [comment]
I've improved the test and I'm iterating locally on a way to propose this.

**Comment by zyga** [comment]
Oh boy, so I made the test reliable but it wasn't pleasant.

**Comment by zyga** [comment]
So this still doesn't pass tests reliably. What is puzzling is that we are getting denials for what is clearly allowed by the profile. It doesn't happen all the time but a simple loop running this single test on --repeat=100 is bound to hit the error.

```
2019-02-02 11:32:15 Error executing google:ubuntu-18.04-64:tests/regression/lp-1813963 : 
-----
+ set +e
+ test-snapd-sh -c /bin/true
+ sleep 31
sanity timeout expired: Interrupted system call
+ exec /home/gopath/src/github.com/snapcore/snapd/tests/regression/lp-1813963/real-snap-update-ns --from-snap-confine test-snapd-sh
cannot send command 1 to helper process: Broken pipe
+ retcode=1
+ set -e
+ test 1 -eq 1
+ set +e
+ systemctl start --wait snap.test-snapd-simple-service.test-snapd-simple-service.service
+ retcode=1
+ set -e
+ test 1 -eq 1
++ grep -c DENIED
++ dmesg
+ test 3 -eq 0
-----
.
2019-02-02 11:32:16 Debug output for google:ubuntu-18.04-64:tests/regression/lp-1813963 : 
-----
+ echo 'Status of the test service'
Status of the test service
+ systemctl status snap.test-snapd-simple-service.test-snapd-simple-service.service
● snap.test-snapd-simple-service.test-snapd-simple-service.service - Service for snap application test-snapd-simple-service.test-snapd-simple-service
   Loaded: loaded (/etc/systemd/system/snap.test-snapd-simple-service.test-snapd-simple-service.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Sat 2019-02-02 10:32:15 UTC; 467ms ago
  Process: 23703 ExecStart=/usr/bin/snap run test-snapd-simple-service (code=exited, status=1/FAILURE)
 Main PID: 23703 (code=exited, status=1/FAILURE)

Feb 02 10:31:44 feb021019-405217 systemd[1]: Started Service for snap application test-snapd-simple-service.test-snapd-simple-service.
Feb 02 10:31:44 feb021019-405217 test-snapd-simple-service.test-snapd-simple-service[23703]: + sleep 31
Feb 02 10:32:14 feb021019-405217 test-snapd-simple-service.test-snapd-simple-service[23703]: sanity timeout expired: Interrupted system call
Feb 02 10:32:15 feb021019-405217 test-snapd-simple-service.test-snapd-simple-service[23703]: + exec /home/gopath/src/github.com/snapcore/snapd/tests/regression/lp-1813963/real-snap-update-ns --from-snap-confine test-snapd-simple-service
Feb 02 10:32:15 feb021019-405217 test-snapd-simple-service.test-snapd-simple-service[23703]: cannot send command 1 to helper process: Broken pipe
Feb 02 10:32:15 feb021019-405217 systemd[1]: snap.test-snapd-simple-service.test-snapd-simple-service.service: Main process exited, code=exited, status=1/FAILURE
Feb 02 10:32:15 feb021019-405217 systemd[1]: snap.test-snapd-simple-service.test-snapd-simple-service.service: Failed with result 'exit-code'.
+ true
+ echo 'Apparmor denials'
Apparmor denials
+ grep DENIED
+ dmesg
[  705.172231] audit: type=1400 audit(1549103473.183:420): apparmor="DENIED" operation="signal" profile="snap-update-ns.test-snapd-simple-service" pid=1 comm="systemd" requested_mask="receive" denied_mask="receive" signal=term peer="unconfined"
[  705.173770] audit: type=1400 audit(1549103473.187:421): apparmor="DENIED" operation="signal" profile="snap-update-ns.test-snapd-simple-service" pid=1 comm="systemd" requested_mask="receive" denied_mask="receive" signal=term peer="unconfined"
[  705.173843] audit: type=1400 audit(1549103473.187:422): apparmor="DENIED" operation="signal" profile="snap-update-ns.test-snapd-simple-service" pid=1 comm="systemd" requested_mask="receive" denied_mask="receive" signal=kill peer="unconfined"
+ '[' 0 = 1 ']'
-----
.
2019-02-02 11:32:16 Starting shell to debug...
google:ubuntu-18.04-64 .../tests/regression/lp-1813963# 
```

As you can see we get a denial for a signal sent by systems to snap-update-ns. This is puzzling because of the profile extension we provide:

```
google:ubuntu-18.04-64 .../tests/regression/lp-1813963# tail /var/lib/snapd/apparmor/profiles/snap-update-ns.test-snapd-simple-service -n 15
  # Don't allow bind mounts to /media which has special
  # sharing and propagates mount events outside of the snap namespace.
  audit deny mount -> /media,

  # Commonly needed permissions for writable mimics.
  /tmp/ r,
  /tmp/.snap/{,**} rw,


/bin/dash mrix,
/bin/sleep mrix,
/home/gopath/src/github.com/snapcore/snapd/tests/regression/lp-1813963/real-snap-update-ns mrix,
signal (send, receive) set=(int, pipe) peer=snap-update-ns.test-snapd-simple-service,
signal (receive) set=(term, kill) peer="unconfined",
}
google:ubuntu-18.04-64 .../tests/regression/lp-1813963# 
```

The profile is also really loaded. If you attempt to load it again the message from the kernel will say so.

**Comment by zyga** [comment]
Just as I wrote this I realised there may be a weird way that `apparmor_parser` is processing the unconfined rule. In all of snapd we say `peer=unconfined` but here I said `peer="unconfined"`. I'll try without the quotes to see if that makes any difference.

**Comment by zyga** [comment]
There's no difference between `peer=unconfined` and `peer="unconfined"`

**Comment by zyga** [comment]
The denial is actually logged not in the main part of the test but in the prepare section. This is odd because snap-update-ns apparmor profile allows all signals from unconfined world (here we get signalled by systemd).

**Comment by zyga** [comment]
Whee, I found the issue :)

**Comment by zyga** [comment]
TL;DR; the error is coming from an unrelated bug in snap-update-ns template that doesn't have the generic signal rule from unconfined world. I misread the source code and though it did. With one extra line in `templates.go` things are running smoothly. I've started a loop of 100 iterations, fingers crossed that it passes.

**Comment by zyga** [comment]
I fixed the test on systems without apparmor support, it should pass now.

**Comment by alfonsosanchezbeato** [comment]
FTR, maybe something interesting for the future would be use `Type=notify` in the service unit, and send a notification to systemd right before the real service is started. This way systemd would not consider the service started until that moment.

**Comment by zyga** [comment]
@alfonsosanchezbeato using Type=notify in snapd is complex https://forum.snapcraft.io/t/its-a-little-bit-hard-to-use-daemon-notify-for-sd-notify/6366

In addition we cannot hijack it as a general feature that all snap services use as a part of their startup. We could craft a new test with daemon type=notify for this specific regression but I don't think it changes anything.

File: cmd/libsnap-confine-private/tool.c [hunk 1]
```diff
@@ -167,7 +167,10 @@ static int sc_open_snapd_tool(const char *tool_name)
 		die("cannot open path %s", dir_name);
 	}
 	int tool_fd = -1;
-	tool_fd = openat(dir_fd, tool_name, O_PATH | O_NOFOLLOW | O_CLOEXEC);
+	/* Use O_CLOEXEC but not O_PATH. This allows us to determine if the file is
+	 * a binary or a script just prior to passing it to fexecve. This way we
+	 * can still undo the use of O_CLOEXEC which is a safe default. */
+	tool_fd = openat(dir_fd, tool_name, O_RDONLY | O_NOFOLLOW | O_CLOEXEC);
```
Comment (mvo5) [inline cmd/libsnap-confine-private/tool.c]: I feel uneasy that we have to modify the code to accommodate for a test in this way. Maybe we can modify the test a bit, something like: http://paste.ubuntu.com/p/Z2k4z2rdCh/  ? I.e. do not use shell for the test but just build a tiny go binary and just run the test on systems that have go (or alternatively install go as a snap where its needed and skip the test on ubuntu-core systems only) or alternatively put a binary that sleeps into a snap and copy that in place if we want the test to run everywhere (but given that is has nothing core or classic specific it seems to be ok to run it on a subset of the systems only?).

File: cmd/snap-confine/ns-support.c [hunk 2]
```diff
@@ -767,6 +787,7 @@ static void sc_wait_for_capture_helper(struct sc_mount_ns *group)
 	}
 	debug("helper process exited normally");
 	group->child = 0;
+	teardown_signals_for_helper();
```
[the remark this answers is not in the document]
Comment (jdstrand) [reply cmd/snap-confine/ns-support.c]: I misread where the teardown was placed; it is down in sc_wait_for_capture_helper() so my analysis was correct and the code is correct in that the parent needs the adjustment for the SIGPIPE. Sorry about misreading that.

Couple of things:
* I think that setup_signals_for_helper() need only be in the parent, no?
* The description focuses a lot on SIGCHLD when it should instead be talking more about SIGPIPE (why are we talking about restarting syscalls when it is simply that the child terminated from SIGALRM?)
* More code comments on why we need to alter SIGPIPE would be helpful
* Please investigate the read() returning 0.

File: cmd/snap-confine/ns-support.c [hunk 3]
```diff
@@ -574,8 +574,28 @@ int sc_join_preserved_per_user_ns(struct sc_mount_ns *group,
 	return ESRCH;
 }
 
+static void setup_signals_for_helper(void)
+{
+	/* Ignore the SIGPIPE signal so that we get EPIPE on the read / write
+	 * operations attempting to work with a closed pipe. This ensures that we
+	 * are not killed by the default disposition (terminate) and can return a
+	 * non-signal-death return code to the program invoking snap-confine. */
+	if (signal(SIGPIPE, SIG_IGN) == SIG_ERR) {
```
[the remark this answers is not in the document]
Comment (mvo5) [reply cmd/snap-confine/ns-support.c]: This use seems to be ok, yes? From the man-page:
```
   Portability
       The only portable use of signal() is to set a signal's  disposition  to
       SIG_DFL  or  SIG_IGN.  The semantics when using signal() to establish a
       signal handler vary across systems (and POSIX.1 explicitly permits this
       variation); do not use it for this purpose.
```

File: cmd/snap-confine/ns-support.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (mvo5) [reply cmd/snap-confine/ns-support.c]: * setup_signals_for_helper(): I looked into this and played with it and I agree that we need the setup_signal_for_helper() only in the parent. I also tested this empirically by setting the signal helper only in the parent and sending a SIGPIPE to the helper. This does not change the exit code of the parent.
* use sigaction instead of signal: see above, AFAICT the usage of signal() in our case is ok
* read() returning 0: good point, I changed the code to to deal with that (and die() with a different error)

File: cmd/snap-confine/ns-support.c [hunk 4]
```diff
@@ -574,6 +574,25 @@ int sc_join_preserved_per_user_ns(struct sc_mount_ns *group,
 	return ESRCH;
 }
 
+static void setup_signals_for_helper(void)
+{
+	/* Ignore the SIGPIPE signal so that we get EPIPE on the read / write
+	 * operations attempting to work with a closed pipe. This ensures that we
+	 * are not killed by the default disposition (terminate) and can return a
+	 * non-signal-death return code to the program invoking snap-confine. */
```
Comment (pedronis) [inline cmd/snap-confine/ns-support.c]: this comment should mention that the issue is that systemd interprets SIGPIPE as a clean termination which is problematic for services that will then not be retried

File: cmd/snap-confine/ns-support.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply cmd/snap-confine/ns-support.c]: we need to rework/simplify the description/commit message of of the PR before landing, that I think what jdstrand was referring to

File: cmd/snap-confine/ns-support.c (hunk 4 above)
Comment (pedronis) [reply cmd/snap-confine/ns-support.c]: this can be done in a follow up

**Comment by pedronis** [comment]
description looks ok now

File: cmd/snap-confine/ns-support.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (jdstrand) [reply cmd/snap-confine/ns-support.c]: Yes, don't worry about sigaction.

File: cmd/snap-confine/ns-support.c (hunk 4 above)
Comment (jdstrand) [reply cmd/snap-confine/ns-support.c]: I agree in a followup is fine (cc @mvo5 and @zyga). That followup should remove SIGPIPE on read though since it is only sent on write (we now properly handle read via @mvo5's changes to this PR).
